---
title: "MFC DESeq analysis"
output:
  html_document: default
  pdf_document: default
---
```{r}
rm(list=ls(all=TRUE)) # clear workspace
graphics.off() # closes all graphics
```

Thanks due to https://huntsmancancerinstitute.github.io/hciR/pasilla_DESeq.html
```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, comment="# ", collapse=TRUE)
# Install function for needed packages ---------------------------------------------------------    
packages<-function(x){
  x<-as.character(match.call()[[2]])
  if (!require(x,character.only=TRUE)){
    install.packages(pkgs=x,repos="http://cran.r-project.org")
    require(x,character.only=TRUE)
  }
}
packages(dplyr)
packages(hciR)
packages(DT)
packages(pasilla)



```

This guide follows the [Bioconductor RNA-Seq workflow] to find differentially expressed
genes using [DESeq2].  Load the [hciR] package to run the [R] code.

### Load samples and counts

Load the counts and samples using the [readr] package.


```{r load_samples, message=FALSE, echo=2:6}
options(width=120)
packages(readr)

packages(microbiome)
samples <- read_tsv("C:/Users/Ryan/Google Drive/UC Davis/Publications/MFC Microbiome Paper - Experiment 15/metadata_frd_exp15.txt")
counts  <- read_tsv("metagenome.tab", quote = "")

# get rid of KEGG_Des
gene_name_lookup <- cbind(counts$OTU_ID,counts$KEGG_Description)
counts <- counts[,-ncol(counts)]
head(counts)
```
Remove features with zero counts and features with one or fewer reads in any sample.

```{r filter_counts}
packages(dplyr)
counts <- filter_counts(counts)
head(counts)
```
Filter data for classes of interest
```{r}
which.cols <- unlist(samples[which(samples$Sample == "Bristles"|samples$Sample == "Cathode"),"Label"])
counts <- as_tibble(cbind(counts$OTU_ID,counts[,which(colnames(counts) %in% which.cols)]))
names(counts)[names(counts) == 'counts$OTU_ID'] <- 'OTU_ID'
counts$OTU_ID <- as.character(counts$OTU_ID)
head(counts)

samples <- samples[which(samples$Sample == "Bristles"|samples$Sample == "Cathode"),]
samples
```



### Run DESeq

Combine the counts and samples to create a `DESeqDataSet` object and calculate
the regularized log transform (rlog) for sample visualizations.

```{r DESeqobject, message=FALSE, warning=FALSE}
packages(lazyeval)
packages(DESeq2)

names(samples)[names(samples) == 'Sample'] <- 'condition'

dds <- deseq_from_tibble(counts, samples,  design = ~ condition)
rld <- r_log(dds)
```

### PCA plot

Plot the first two principal components using the rlog values from the top 500 variable genes.
You can hover over points to view sample names or zoom into groups of points in this interactive [highchart].

```{r pcaplot, fig.height=7,fig.width=7}
plot_pca(rld, "condition", tooltip=c("Label") , width=700)
```

Cluster all the rlog values using the R function `dist` to calculate the Euclidean distance between samples.

```{r dist}
plot_dist(rld , c("condition"), palette="Reds", diagNA=FALSE, reverse_pal=FALSE)
```

### Gene annotations

Load/create annotations.

```{r biomaRt}
fly <- as_tibble(gene_name_lookup)
colnames(fly) <- c("id","gene_name")
fly <- fly[order(fly$id),]
#add some NA columns
namevector <- c("biotype", "chromosome", "description")
fly[ , namevector] <- NA

head(fly)
```

### Result tables

Get the annotated DESeq results using a 5% false discovery rate (FDR).  The default is to compare
all treatment combinations, which only includes one in this study.

Since we downloaded the most recent fly annotations, there are 723 missing (note you could set the version
in read_biomart to match the old count table)


```{r DESeq_results}
res <- results_all(dds, fly, alpha= 0.05)
```
### Browse results

Create a [flex dashboard] using the top 2000 genes sorted by p-value in
[pasilla_flex.html].   The [MA-plot] and [volcano plot] are linked, so you can
click and drag to create a box to highlight points and then view matching rows
in the table. You can also drag the box around the plot to easily highlight
other points and rows. In addition, you can search for genes in the table using
the search box, but need to click on the row in order to highlight the points in
the plots.  The sliders can also be used to limit the results.

```{r flex_dash, message=FALSE, eval=FALSE}
# packages(rmarkdown)
# packages(flexdashboard)
# packages(crosstalk)
# packages(d3scatter)
# rmd <- system.file("Rmd", "DESeq_flex.Rmd", package="hciR")
# render(rmd, output_file="pasilla_flex.html",  output_dir=".",
#          params=list( results= res, title = "Treatments", top= 2000 ))
```

### Save results

Save the DESeq results to a single Excel file in [pasilla_DESeq.xlsx].  The `write_deseq` function will also
output raw counts, rlog values, normalized counts, samples and fly annotations.

```{r write, eval=FALSE}
# write_deseq(res, dds, rld, fly, file="pasilla_DESeq.xlsx")
```
### Plots

Plot the fold changes and p-values in an interactive volcano plot.

```{r volcano}
plot_volcano(res, padj = 1e-10, log2Fold =1 , width=700)
```

<br>

Select the top 40 genes sorted by p-value and cluster the rlog differences,
so values in the heatmap represent the amount a gene deviates in a specific sample
from the geneâ€™s average across all samples.

<br>

```{r plot1, fig.width=8, fig.height=7}
x <- top_counts( res, rld, top=40)
x
plot_genes(x, c("condition") )
```
<br>

<P style="page-break-before: always">
Plot the top 400 genes using an interactive [d3heatmap].  Click and drag over a
region in the plot to zoom and better view gene labels.

```{r plotd3, fig.width=8, fig.height=9}
plot_genes( top_counts( res, rld, top=400), output="d3", xaxis_font_size=12, show_grid=FALSE)
```

<br>
<br>


[readr]: http://readr.tidyverse.org/
[highchart]: http://jkunst.com/highcharter/index.html
[pasilla]: http://bioconductor.org/packages/release/data/experiment/html/pasilla.html
[R for Data Science]: http://r4ds.had.co.nz/
[R]: https://www.r-project.org
[Bioconductor RNA-Seq workflow]: http://www.bioconductor.org/help/workflows/rnaseqGene
[DESeq2]: http://www.bioconductor.org/packages/release/bioc/html/DESeq2.html
[hciR]: https://github.com/HuntsmanCancerInstitute/hciR
[pheatmap]: https://cran.r-project.org/web/packages/pheatmap/index.html
[d3heatmap]: http://www.htmlwidgets.org/showcase_d3heatmap.html

[flex dashboard]: http://rmarkdown.rstudio.com/flexdashboard/
[MA-plot]: https://en.wikipedia.org/wiki/MA_plot
[volcano plot]: https://en.wikipedia.org/wiki/Volcano_plot_(statistics)
[pasilla_flex.html]: https://huntsmancancerinstitute.github.io/hciR/pasilla_flex.html
[pasilla_DESeq.xlsx]: https://huntsmancancerinstitute.github.io/hciR/pasilla_DESeq.xlsx
[pasilla_DESeq.Rmd]: https://github.com/HuntsmanCancerInstitute/hciR/blob/master/inst/Rmd/pasilla_DESeq.Rmd
